src/viz/viewer_renderer.cpp

```diff
@@ -416,6 +416,227 @@ static VkQueue makeGFXQueue(const Device &dev, uint32_t idx)
+        std::move(shader),
+            voxel_gen_layout,
+            pipelines,
+            std::move(desc_pool),
+    };
+}
+
+static PipelineShaders makeVoxelDrawShaders(
+    const Device& dev, VkSampler repeat_sampler, VkSampler clamp_sampler)
+{
+    (void)repeat_sampler;
+    (void)clamp_sampler;
+
+    std::filesystem::path shader_dir =
+        std::filesystem::path(STRINGIFY(VIEWER_DATA_DIR)) /
+        "shaders";
+
+    auto shader_path = (shader_dir / "voxel_draw.hlsl").string();
+
+    ShaderCompiler compiler;
+    SPIRVShader vert_spirv = compiler.compileHLSLFileToSPV(
+        shader_path.c_str(), {}, {},
+        { "vert", ShaderStage::Vertex });
+
+    SPIRVShader frag_spirv = compiler.compileHLSLFileToSPV(
+        shader_path.c_str(), {}, {},
+        { "frag", ShaderStage::Fragment });
+        std::array<SPIRVShader, 2> shaders {
+        std::move(vert_spirv),
+            std::move(frag_spirv),
+    };
+
+    StackAlloc tmp_alloc;
+    return PipelineShaders(dev, tmp_alloc, shaders,
+        Span<const BindingOverride>({
+            BindingOverride {
+                1,
+                0,
+                VK_NULL_HANDLE,
+                InternalConfig::maxTextures,
+                VK_DESCRIPTOR_BINDING_PARTIALLY_BOUND_BIT,
+            },
+            BindingOverride {
+                1,
+                1,
+                repeat_sampler,
+                1,
+                0,
+            },
+            }));
+}
+
+static void issueVoxelGen(Device& dev,
+    Frame& frame,
+    Pipeline<1>& pipeline,
+    VkCommandBuffer draw_cmd,
+    uint32_t view_idx,
+    uint32_t max_views,
+    VoxelConfig voxel_config)
+{
+    const uint32_t num_voxels = voxel_config.xLength * voxel_config.yLength * voxel_config.zLength;
+
+    {
+        VkBufferMemoryBarrier compute_prepare = {
+            VK_STRUCTURE_TYPE_BUFFER_MEMORY_BARRIER,
+            nullptr,
+           VK_ACCESS_MEMORY_WRITE_BIT | VK_ACCESS_MEMORY_READ_BIT,
+            VK_ACCESS_MEMORY_WRITE_BIT|VK_ACCESS_MEMORY_READ_BIT,
+            VK_QUEUE_FAMILY_IGNORED, VK_QUEUE_FAMILY_IGNORED,
+            frame.voxelData.buffer,
+            0,
+            (VkDeviceSize)(num_voxels * sizeof(uint32_t))
+        };
+
+        dev.dt.cmdPipelineBarrier(draw_cmd, VK_PIPELINE_STAGE_TRANSFER_BIT, VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT,
+            0, 0, nullptr, 1, &compute_prepare, 0, nullptr);
+    }
+
+    {
+        VkBufferMemoryBarrier compute_prepare = {
+            VK_STRUCTURE_TYPE_BUFFER_MEMORY_BARRIER,
+            nullptr,
+           VK_ACCESS_MEMORY_WRITE_BIT | VK_ACCESS_MEMORY_READ_BIT,
+            VK_ACCESS_MEMORY_WRITE_BIT | VK_ACCESS_MEMORY_READ_BIT,
+            VK_QUEUE_FAMILY_IGNORED, VK_QUEUE_FAMILY_IGNORED,
+            frame.voxelVBO.buffer,
+            0,
+            (VkDeviceSize)(32 * 4 * 6 * num_voxels)
+        };
+
+        dev.dt.cmdPipelineBarrier(draw_cmd, VK_PIPELINE_STAGE_TRANSFER_BIT, VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT,
```

Comment:
Code base has a 80 char width max unless there is a serious reason to go longer (the renderer may not be consistent about this due to Luc copying in some code).